File: Iterator/Dart/UserCollection.dart
import 'ICollection.dart';
import 'IIterator.dart';
import 'User.dart';
import 'UserIterator.dart';

class UserCollection implements ICollection {
  List<User> _users = [];

  @override
  IIterator createIterator() {
    return UserIterator(_users);
  }

  add(User user) {
    _users.add(user);
  }
}

File: Iterator/Javascript/UserCollection.js
import ICollection from "./ICollection.js";
import UserIterator from "./UserIterator.js";

class UserCollection extends ICollection {
  constructor() {
    super();
    this.users = [];
  }

  createIterator() {
    return new UserIterator(this.users);
  }

  add(user) {
    this.users.push(user);
  }
}

export default UserCollection;